### DISCUSSION TITLE
Is there a plan to support ViewTransition API?

### DISCUSSION THREAD

**Discussion by push1kb** [TanStack/router#1366] [Q&A]
I did not see anything related to that issues or pull requests. Is that in roadmap, or ignored for now?

Thanks in advance

**Comment by SeanCassiere** [comment]
Out of curiosity, what's missing for you to be able to implement view transitions?

The `<Link>` component allows you to add an `onClick` callback, and the `Router` has a `.subscribe` method that allows you to watch for events (like when navigation changes).

Outside of this, is there anything missing?

Edit: Found [this article](https://malcolmkee.com/blog/view-transition-api-in-react-app/) on implementing view transitions in React, and from the outskirts it looks like something that should be possible right?

**Reply by push1kb** [reply]
Thanks for providing me that article. I managed to implement it via `onClick` callback

**Reply by schiller-manuel** [reply]
would be cool to see a demo, can you share a repo?

**Comment by flipvh** [comment]
I am also interested in this. Anybody stumbled upon this discussion and has an example that can be shared? Will work on it myself soon, so if not, will be sharing progress here.

**Reply by tannerlinsley** [reply]
Working on it right now. Hang tight :)

**Reply by Flam9** [reply]
+1 also interested in this. Saw this demo that was mindblowingly cool https://live-transitions.pages.dev/

**Reply by jeremyosih** [reply]
+1 would be awesome to have that implement :) But understand you have a lot going on with Start, just showing my interest here :)

**Reply by SeanCassiere** [reply]
We do support view transitions and can be enabled on the router using `defaultViewTransition: true`.
https://tanstack.com/router/latest/docs/framework/react/api/router/RouterOptionsType#defaultviewtransition-property

**Reply by aretrace** [reply]
❤️

**Reply by jeremyosih** [reply]
Thank you so much ! :)

**Reply by tigawanna** [reply]
will it becalled automatically on route change and how do i define custo  transitions?

**Comment by asnaeb** [comment]
how about viewTransition types? This doesn't seem to be feasible with the current api.

https://developer.chrome.com/docs/web-platform/view-transitions/same-document#view-transition-types

I took a quick look at the source code and seems like it can be supported with little effort. @SeanCassiere Would a PR to support this be welcome?

**Reply by SeanCassiere** [reply]
Would definitely be open to PRs here.

My only feedback here is that the API would need to support something that's applicable on general navigations (like using `<Link>` and `useNavigate() -> navigate(...)`) as well as being able to add these settings on the `createRouter` function as defaults.

**Reply by asnaeb** [reply]
Okay, I'll be looking into it as soon as I can

**Reply by functionsandglory** [reply]
@asnaeb @SeanCassiere  I see that default types can be added to `defaultViewTransition` router prop (not sure if your work).

But I was wondering if it is possible to dynamically add `types` per route and more dynamically to accomplish something like found here:

https://developer.chrome.com/docs/web-platform/view-transitions/same-document#view-transition-types

**Reply by functionsandglory** [reply]
> @asnaeb @SeanCassiere I see that default types can be added to `defaultViewTransition` router prop (not sure if your work).
> 
> But I was wondering if it is possible to dynamically add `types` per route and more dynamically to accomplish something like found here:
> 
> https://developer.chrome.com/docs/web-platform/view-transitions/same-document#view-transition-types

Nevermind, I see it here: https://github.com/TanStack/router/commit/70df20a53caa12be84e378a115b6247cee798cb7#diff-5318334b218ab6668c835a42a6cef2f9e0e390764e92204674fd8b26b4e52f1a

Just needs better documentation. I can maybe help with that.
